Docs: Contributing folder (#4415)
langflow-ai/langflow: docs/docs/Contributing/contributing-how-to-contribute.md

slug: /contributing-how-to-contribute
---

> # How to contribute to Langflow

Reviewer: I'm not sure why you have this heading.

> This guide is intended to help you get started contributing to Langflow.

Reviewer: ```suggestion
This guide is intended to help you start contributing to Langflow.
```

> As an open-source project in a rapidly developing field, we are extremely open
> to contributions, whether it be in the form of a new feature, improved infra, or better documentation.

Reviewer: ```suggestion
As an open-source project in a rapidly developing field, Langflow welcomes contributions, whether it be in the form of a new feature, improved infrastructure, or better documentation.
```

> To contribute to this project, please follow the [fork and pull request](https://docs.github.com/en/get-started/quickstart/contributing-to-projects) workflow.

Reviewer: ```suggestion
To contribute to this project, follow the [fork and pull request](https://docs.github.com/en/get-started/quickstart/contributing-to-projects) workflow.
```

> To contribute to this project, please follow the [fork and pull request](https://docs.github.com/en/get-started/quickstart/contributing-to-projects) workflow.

Reviewer: Does this sentence apply to both contributing to code and to docs? It's not clear.

Author: Clarified

> ## Prerequisites

Reviewer: Do these prerequisites apply to both contributing to code and to docs? It's not clear.

Author: Good call, split them

- [uv(>=0.4)](https://docs.astral.sh/uv/getting-started/installation/)
- [Node.js](https://nodejs.org/en/download/package-manager)

## Contribute code

Develop Langflow locally with [uv](https://docs.astral.sh/uv/getting-started/installation/) and [Node.js](https://nodejs.org/en/download/package-manager).

### Clone the Langflow Repository

> 1. Navigate to the [Langflow GitHub repository](https://github.com/langflow-ai/langflow) and click **Fork**.

Reviewer: ```suggestion
1. Navigate to the [Langflow GitHub repository](https://github.com/langflow-ai/langflow), and then click **Fork**.
```

2. Add the new remote to your local repository on your local machine:

```bash
git remote add fork https://github.com/<your_git_username>/langflow.git
```

### Prepare the development environment

1. Create development hooks.

```bash
make init
```

This command sets up the development environment by installing backend and frontend dependencies, building the frontend static files, and initializing the project. It runs `make install_backend`, `make install_frontend`, `make build_frontend`, and finally `uv run langflow run` to start the application.

2. Run `make lint`, `make format`, and `make unit_tests` before pushing to the repository.

### Debug

> The repo includes a `.vscode/launch.json` file for debugging the backend in VSCode, which is a lot faster than debugging with Docker compose. To debug Langflow with the `launch.json` file in VSCode:

Reviewer: ```suggestion
The repo includes a `.vscode/launch.json` file for debugging the backend in VSCode, which is faster than debugging with Docker Compose. To debug Langflow with the `launch.json` file in VSCode:
```

1. Open Langflow in VSCode.

> 2. Press Ctrl+Shift+D (or Cmd+Shift+D on Mac) to open the Run and Debug view.

Reviewer: ```suggestion
2. Press **Ctrl+Shift+D** for Windows **or Cmd+Shift+D** for Mac to open the Run and Debug view.
```

> 3. Choose a configuration from the dropdown at the top (for example, "Debug Backend").

Reviewer: Two things here. First, "from the dropdown at the top" is vague. Does the dropdown have a label? Second, always put the location before the action, so it should be "From the dropdown at the top, choose a configuration"

4. Click the green play button or press F5 to start debugging.

> 5. This allows you to quickly debug different parts of your application, like the backend, frontend, or CLI, directly from VSCode.

Reviewer: This is informational, not an action that has to be taken, so it should not be a numbered step.

> ### Run Langflow locally (Poetry and Node.js)

Reviewer: Why are Poetry and Node.js in parentheses?

> Run Langflow locally by cloning the repository and installing the dependencies. We recommend using a virtual environment like venv or conda to isolate dependencies.

Reviewer: Avoid the use of "we". Use "Langflow" instead.

> Run Langflow locally by cloning the repository and installing the dependencies. We recommend using a virtual environment like venv or conda to isolate dependencies.

Reviewer: Are you telling users to run Langflow locally so they can test their changes? It's not clear.

Author: Added more explanation

Before you begin, ensure you have [uv](https://docs.astral.sh/uv/getting-started/installation/) and [Node.js](https://nodejs.org/en/download/package-manager) installed.

1. In the repository root, install the dependencies and start the development server for the backend:

```bash
make backend
```

2. Install dependencies and start the frontend:

```bash
make frontend
```

## Contribute documentation

> The documentation is built using [Docusaurus](https://docusaurus.io/). To run the documentation locally, run the following commands:

Reviewer: You should also mention Markdown.

```bash
cd docs
npm install
npm run start
```

The documentation will be available at `localhost:3000` and all the files are located in the `docs/docs` folder.

## Open a pull request

> Once you wrote and manually tested your change, you can start sending the patch to the main repository.

Reviewer: ```suggestion
Once you have written and manually tested your change, you can start sending the patch to the main repository.
```

- Open a new GitHub pull request with the patch against the `main` branch.

> - Ensure the PR title follows semantic commits conventions.

Reviewer: ```suggestion
- Ensure the PR title follows semantic commit conventions.
```

## Open a pull request

Once you have written and manually tested your changes with `make lint` and `make unit_tests`, open a pull request to send your changes upstream to the main Langflow repository.

1. Open a new GitHub pull request with your patch against the `main` branch.
2. Ensure the PR title follows semantic commits conventions. For example, features are `feat: add new feature` and fixes are `fix: correct issue with X`.
3. A Langflow maintainer will review your pull request. Thanks for your contribution!

Some additional guidance on pull request titles:
* Ensure the pull request description clearly describes the problem and solution. If the PR fixes an issue, include a link to the fixed issue in the PR description with `Fixes #1234`.
* Pull request titles appear in Langflow's release notes, so they should explain what the PR does as explicitly as possible.
* Pull requests should strive to fix one thing **only**, and should contain a good description of what is being fixed.

For more information, see the [Python Developer's Guide](https://devguide.python.org/getting-started/pull-request-lifecycle/index.html#making-good-commits).

> ## Report bugs or suggest improvements

Reviewer: Was is an oversight that this information is here and in a separate topic? I'm not sure why you have it in two places.
